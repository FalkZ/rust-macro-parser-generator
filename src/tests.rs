macro_rules! tests {
    (@debug: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_debug_snapshot!(&r.$t);
        }
    };
    (@display: $name:ident, $t:ident, $str:literal) => {
        #[test]
        fn $name() {
            let r = crate::m1n::renderer::render($str).expect("couldn't compile file");
            insta::assert_display_snapshot!(&r.$t);
        }
    };
    // `()` indicates that the macro takes no argument.
    ($($name:ident: $str:literal),*) => {
        mod tokens {
            $(
                tests!(@debug: $name, tokens, $str);
            )*
        }
        mod statements {
            $(
                tests!(@debug: $name, statements, $str);
            )*
        }
        mod typescript {
            $(
                tests!(@display: $name, typescript, $str);
            )*
        }
    };
}

tests!(
    enums: "./examples/m1n/Enum.m1n",
    class: "./examples/m1n/Class.m1n",
    file: "./examples/m1n/file.m1n",
    math: "./examples/m1n/std/math.m1n",
    matcher: "./examples/m1n/match.m1n",
    assignment: "./examples/m1n/assignment.m1n",
    operator: "./examples/m1n/operator.m1n"
);
