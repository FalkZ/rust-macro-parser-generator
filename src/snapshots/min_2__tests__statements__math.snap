---
source: src/tests.rs
expression: "&r.statements"
---
[
    statements {
        position: Position {
            column: 0,
            line: 1,
        },
        statement: imports(
            imports {
                position: Position {
                    column: 0,
                    line: 1,
                },
                imports: [
                    import {
                        position: Position {
                            column: 4,
                            line: 2,
                        },
                        path: RAWIDENT(
                            Token {
                                raw: "`@std/util`",
                                position: Position {
                                    column: 4,
                                    line: 2,
                                },
                            },
                        ),
                        import_items: Some(
                            import_items {
                                position: Position {
                                    column: 15,
                                    line: 2,
                                },
                                items: [
                                    import_item {
                                        position: Position {
                                            column: 17,
                                            line: 2,
                                        },
                                        item: IDENT(
                                            Token {
                                                raw: "pipe",
                                                position: Position {
                                                    column: 17,
                                                    line: 2,
                                                },
                                            },
                                        ),
                                    },
                                    import_item {
                                        position: Position {
                                            column: 23,
                                            line: 2,
                                        },
                                        item: IDENT(
                                            Token {
                                                raw: "match",
                                                position: Position {
                                                    column: 23,
                                                    line: 2,
                                                },
                                            },
                                        ),
                                    },
                                ],
                            },
                        ),
                    },
                    import {
                        position: Position {
                            column: 4,
                            line: 3,
                        },
                        path: RAWIDENT(
                            Token {
                                raw: "`@std/math`",
                                position: Position {
                                    column: 4,
                                    line: 3,
                                },
                            },
                        ),
                        import_items: None,
                    },
                ],
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 4,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 4,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 4,
                        },
                        modifier: PUB(
                            Token {
                                raw: "pub",
                                position: Position {
                                    column: 0,
                                    line: 4,
                                },
                            },
                        ),
                    },
                ],
                name: RAWIDENT(
                    Token {
                        raw: "`+`",
                        position: Position {
                            column: 4,
                            line: 4,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 8,
                            line: 4,
                        },
                        arguments: [
                            argument {
                                position: Position {
                                    column: 8,
                                    line: 4,
                                },
                                arg: IDENT(
                                    Token {
                                        raw: "a",
                                        position: Position {
                                            column: 8,
                                            line: 4,
                                        },
                                    },
                                ),
                            },
                        ],
                        last: IDENT(
                            Token {
                                raw: "b",
                                position: Position {
                                    column: 11,
                                    line: 4,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 16,
                        line: 4,
                    },
                    value: primitive_value(
                        TYPESCRIPT(
                            Token {
                                raw: "#a + b#",
                                position: Position {
                                    column: 16,
                                    line: 4,
                                },
                            },
                        ),
                    ),
                    expressions: [],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 6,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 6,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 6,
                        },
                        modifier: PUB(
                            Token {
                                raw: "pub",
                                position: Position {
                                    column: 0,
                                    line: 6,
                                },
                            },
                        ),
                    },
                ],
                name: RAWIDENT(
                    Token {
                        raw: "`-`",
                        position: Position {
                            column: 4,
                            line: 6,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 8,
                            line: 6,
                        },
                        arguments: [
                            argument {
                                position: Position {
                                    column: 8,
                                    line: 6,
                                },
                                arg: IDENT(
                                    Token {
                                        raw: "a",
                                        position: Position {
                                            column: 8,
                                            line: 6,
                                        },
                                    },
                                ),
                            },
                        ],
                        last: IDENT(
                            Token {
                                raw: "b",
                                position: Position {
                                    column: 11,
                                    line: 6,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 16,
                        line: 6,
                    },
                    value: primitive_value(
                        TYPESCRIPT(
                            Token {
                                raw: "#a - b#",
                                position: Position {
                                    column: 16,
                                    line: 6,
                                },
                            },
                        ),
                    ),
                    expressions: [],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 8,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 8,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 8,
                        },
                        modifier: PUB(
                            Token {
                                raw: "pub",
                                position: Position {
                                    column: 0,
                                    line: 8,
                                },
                            },
                        ),
                    },
                ],
                name: RAWIDENT(
                    Token {
                        raw: "`*`",
                        position: Position {
                            column: 4,
                            line: 8,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 8,
                            line: 8,
                        },
                        arguments: [
                            argument {
                                position: Position {
                                    column: 8,
                                    line: 8,
                                },
                                arg: IDENT(
                                    Token {
                                        raw: "a",
                                        position: Position {
                                            column: 8,
                                            line: 8,
                                        },
                                    },
                                ),
                            },
                        ],
                        last: IDENT(
                            Token {
                                raw: "b",
                                position: Position {
                                    column: 11,
                                    line: 8,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 16,
                        line: 8,
                    },
                    value: primitive_value(
                        TYPESCRIPT(
                            Token {
                                raw: "#a * b#",
                                position: Position {
                                    column: 16,
                                    line: 8,
                                },
                            },
                        ),
                    ),
                    expressions: [],
                },
            },
        ),
    },
    statements {
        position: Position {
            column: 0,
            line: 10,
        },
        statement: function(
            function {
                position: Position {
                    column: 0,
                    line: 10,
                },
                modifiers: [
                    modifiers {
                        position: Position {
                            column: 0,
                            line: 10,
                        },
                        modifier: PUB(
                            Token {
                                raw: "pub",
                                position: Position {
                                    column: 0,
                                    line: 10,
                                },
                            },
                        ),
                    },
                ],
                name: RAWIDENT(
                    Token {
                        raw: "`/`",
                        position: Position {
                            column: 4,
                            line: 10,
                        },
                    },
                ),
                arguments: Some(
                    arguments {
                        position: Position {
                            column: 8,
                            line: 10,
                        },
                        arguments: [
                            argument {
                                position: Position {
                                    column: 8,
                                    line: 10,
                                },
                                arg: IDENT(
                                    Token {
                                        raw: "a",
                                        position: Position {
                                            column: 8,
                                            line: 10,
                                        },
                                    },
                                ),
                            },
                        ],
                        last: IDENT(
                            Token {
                                raw: "b",
                                position: Position {
                                    column: 11,
                                    line: 10,
                                },
                            },
                        ),
                    },
                ),
                body: body {
                    position: Position {
                        column: 16,
                        line: 10,
                    },
                    value: primitive_value(
                        TYPESCRIPT(
                            Token {
                                raw: "#a / b#",
                                position: Position {
                                    column: 16,
                                    line: 10,
                                },
                            },
                        ),
                    ),
                    expressions: [],
                },
            },
        ),
    },
]
